tests/e2e-tests/src/recursor/dnssec/scenarios.rs

```diff
@@ -0,0 +1,242 @@
+                                rrsig.signature = "bm90YXJlYWxycnNpZ3JlY29yZA==".to_string();
+                                *record = Record::RRSIG(rrsig);
+                                did_remove = true;
+                            }
+                        }
+                    }
+                    assert!(
+                        did_remove,
+                        "did not find an RRSIG covering DNSKEY generated using the KSK"
+                    );
+                }
+            },
+        },
+    )?;
+
+    let mut resolver = Resolver::new(&network, root);
+
+    let resolver = resolver
+        .trust_anchor(&trust_anchor.unwrap())
+        .start_with_subject(&Implementation::hickory())?;
+
+    let client = Client::new(resolver.network())?;
+    let settings = *DigSettings::default().recurse().authentic_data();
+    let output = client.dig(
+        settings,
+        resolver.ipv4_addr(),
+        RecordType::A,
+        &leaf_zone.push_label("noexist"),
+    )?;
+
+    let logs = resolver.logs()?;
+
+    dbg!(&output);
+
+    assert!(output.status.is_servfail());
+    assert!(!logs.contains("iteration count 501 is over 500"));
+    assert!(logs.contains("response does not contain NSEC or NSEC3 records."));
+
+    Ok(())
+}
+
+// the `no-ds.testing.` zone is signed but no DS record exists in the parent `testing.` zone.
+// importantly, the `testing.` zone must contain NSEC/NSEC3 records to deny the existence of
+// `no-ds.testing./DS` (which is why we cannot use `Graph::build` + `Sign::AndAmend` to produce
+// this network)
+pub fn insecure_record_fixture(
+    query_fqdn: FQDN,
+    sign_settings: SignSettings,
+    custom_config: Option<String>,
+) -> Result<(DigOutput, String)> {
+    let network = Network::new()?;
+
+    let insecure_zone = FQDN::TEST_TLD.push_label("insecure");
+    let needle_fqdn = FQDN("example.insecure.testing.")?;
+    let needle_ipv4_addr = Ipv4Addr::new(1, 2, 3, 4);
+
+    let mut insecure_ns = NameServer::new(
+        &Implementation::test_peer(),
+        insecure_zone.clone(),
+        &network,
+    )?;
+    insecure_ns.add(Record::a(needle_fqdn.clone(), needle_ipv4_addr));
+
+    let mut sibling_ns =
+        NameServer::new(&Implementation::test_peer(), FQDN::TEST_DOMAIN, &network)?;
+    let mut tld_ns = NameServer::new(&Implementation::test_peer(), FQDN::TEST_TLD, &network)?;
+    let mut root_ns = NameServer::new(&Implementation::test_peer(), FQDN::ROOT, &network)?;
+
+    sibling_ns.add(root_ns.a());
+    sibling_ns.add(tld_ns.a());
+    sibling_ns.add(insecure_ns.a());
+    sibling_ns.add(sibling_ns.a());
+
+    root_ns.referral_nameserver(&tld_ns);
+    tld_ns.referral_nameserver(&sibling_ns);
+    tld_ns.referral_nameserver(&insecure_ns);
+
+    let insecure_ns = insecure_ns.sign(sign_settings.clone())?;
+    let sibling_ns = sibling_ns.sign(sign_settings.clone())?;
+
+    tld_ns.add(sibling_ns.ds().ksk.clone());
+    tld_ns.add(insecure_ns.ds().ksk.clone());
```

Comment:
The comment at the top of this fixture function and the `insecure_ns` variable name are confusing. I think this line means that the `insecure.testing.` zone is in fact a secure zone. Indeed, we'd need it to be a secure zone, so that these tests can trigger it to be treated as an insecure zone by raising the NSEC3 iterations count. I suggest renaming this to `leaf_ns` and deleting the copied comment before the function.